luskan/src/main/java/org/opencadc/luskan/NumCopiesNavigator.java

```diff
@@ -0,0 +1,142 @@
+ *  the GNU Affero General Public        la “GNU Affero General Public
+ *  License as published by the          License” telle que publiée
+ *  Free Software Foundation,            par la Free Software Foundation
+ *  either version 3 of the              : soit la version 3 de cette
+ *  License, or (at your option)         licence, soit (à votre gré)
+ *  any later version.                   toute version ultérieure.
+ *
+ *  OpenCADC is distributed in the       OpenCADC est distribué
+ *  hope that it will be useful,         dans l’espoir qu’il vous
+ *  but WITHOUT ANY WARRANTY;            sera utile, mais SANS AUCUNE
+ *  without even the implied             GARANTIE : sans même la garantie
+ *  warranty of MERCHANTABILITY          implicite de COMMERCIALISABILITÉ
+ *  or FITNESS FOR A PARTICULAR          ni d’ADÉQUATION À UN OBJECTIF
+ *  PURPOSE.  See the GNU Affero         PARTICULIER. Consultez la Licence
+ *  General Public License for           Générale Publique GNU Affero
+ *  more details.                        pour plus de détails.
+ *
+ *  You should have received             Vous devriez avoir reçu une
+ *  a copy of the GNU Affero             copie de la Licence Générale
+ *  General Public License along         Publique GNU Affero avec
+ *  with OpenCADC.  If not, see          OpenCADC ; si ce n’est
+ *  <http://www.gnu.org/licenses/>.      pas le cas, consultez :
+ *                                       <http://www.gnu.org/licenses/>.
+ *
+ *  : 5 $
+ *
+ ************************************************************************
+ */
+
+package org.opencadc.luskan;
+
+import ca.nrc.cadc.tap.parser.ParserUtil;
+import ca.nrc.cadc.tap.parser.navigator.FromItemNavigator;
+import ca.nrc.cadc.tap.parser.navigator.ReferenceNavigator;
+import ca.nrc.cadc.tap.parser.navigator.SelectNavigator;
+import java.util.ArrayList;
+import java.util.List;
+import net.sf.jsqlparser.expression.Expression;
+import net.sf.jsqlparser.schema.Table;
+import net.sf.jsqlparser.statement.select.PlainSelect;
+import net.sf.jsqlparser.statement.select.SelectItem;
+import org.apache.log4j.Logger;
+
+/**
+ * Checks there is a single Artifact table in the from clause,
+ * then applies the NumCopiesConverter to the select, where, and group by clauses.
+ *
+ */
+public class NumCopiesNavigator extends SelectNavigator {
+    private static final Logger log = Logger.getLogger(NumCopiesNavigator.class);
+
+    NumCopiesConverter numCopiesConverter;
+
+    public NumCopiesNavigator(NumCopiesConverter en, ReferenceNavigator rn, FromItemNavigator fn) {
+        super(en, rn, fn);
+        this.numCopiesConverter = en;
+    }
+
+    @Override
+    public void visit(PlainSelect ps) {
+        log.debug("visit(PlainSelect) " + ps);
+
+        List<Table> tables = new ArrayList<>();
+        List<Table> fromTableList = ParserUtil.getFromTableList(ps);
+        for (Table fromTable : fromTableList) {
+            if (fromTable.getName().equalsIgnoreCase("artifact")) {
```

Comment:
There could be a table named `tap_upload.artifact` in the query, so you have to look for `inventory.artifact` (that's the name in the tap_schema, so that's the name queries have to use).

Reply:
The loop is looking for all Artifact tables. If there are multiple Artifact tables the methods exits without doing any work. But there does need to be a check if there's a single table in the FROM, it is an inventory.Artifact table.